enterprise/internal/codeintel/resolvers/location.go

```diff
@@ -50,28 +54,89 @@ func (r *locationConnectionResolver) PageInfo(ctx context.Context) (*graphqlutil
+	for _, location := range locations {
+		if _, ok := pathsByCommitByRepository[location.Repository]; !ok {
+			pathsByCommitByRepository[location.Repository] = pathsByCommitMap{}
+		}
+
+		if _, ok := pathsByCommitByRepository[location.Repository][location.Commit]; !ok {
+			pathsByCommitByRepository[location.Repository][location.Commit] = pathsSet{}
+		}
+
+		pathsByCommitByRepository[location.Repository][location.Commit][location.Path] = struct{}{}
 	}
 
-	commitResolver, err := graphqlbackend.NewRepositoryResolver(repo).Commit(
-		ctx,
-		&graphqlbackend.RepositoryCommitArgs{Rev: location.Commit},
-	)
-	if err != nil {
-		return nil, err
+	return pathsByCommitByRepository
+}
+
+type resolversByPathByCommitByRepostioryMap map[string]resolversByPathByCommitMap
+type resolversByPathByCommitMap map[string]resolversByPathMap
+type resolversByPathMap map[string]*graphqlbackend.GitTreeEntryResolver
+
+// resolveGitTrees takes the map produced by `resolveGitTrees` and returns a symmetric map, where the
+// empty structs in the path set are replaced by git tree resolvers. This ensures that each repository,
+// commit, and path are each resolved only once.
+//
+// This method resolves repositories. See `resolveGitCommitsForRepository` and `resolveGitTreeForPath`
+// (which are called from here) for the resolution of commits and git trees, respectively.
+func resolveGitTrees(ctx context.Context, pathsByCommitByRepository pathsByCommitByRepositoryMap) (resolversByPathByCommitByRepostioryMap, error) {
+	resolversByRepositories := resolversByPathByCommitByRepostioryMap{}
+
+	for repoName, commits := range pathsByCommitByRepository {
+		repo, err := backend.Repos.GetByName(ctx, api.RepoName(repoName))
+		if err != nil {
+			return nil, err
+		}
+		repositoryResolver := graphqlbackend.NewRepositoryResolver(repo)
+
+		resolvers, err := resolveGitCommitsForRepository(ctx, repositoryResolver, commits)
+		if err != nil {
+			return nil, err
+		}
+		resolversByRepositories[repoName] = resolvers
 	}
 
-	gitTreeResolver, err := commitResolver.Blob(ctx, &struct {
-		Path string
-	}{
-		Path: location.Path,
-	})
-	if err != nil {
-		return nil, err
+	return resolversByRepositories, nil
+}
+
+func resolveGitCommitsForRepository(ctx context.Context, repositoryResolver *graphqlbackend.RepositoryResolver, pathsByCommit pathsByCommitMap) (resolversByPathByCommitMap, error) {
+	resolversByCommit := resolversByPathByCommitMap{}
+
+	for commit, paths := range pathsByCommit {
+		commitResolver, err := repositoryResolver.Commit(ctx, &graphqlbackend.RepositoryCommitArgs{Rev: commit})
+		if err != nil {
+			return nil, err
+		}
+
+		resolvers, err := resolveGitTreeForPath(ctx, commitResolver, paths)
+		if err != nil {
+			return nil, err
+		}
+		resolversByCommit[commit] = resolvers
+	}
+
+	return resolversByCommit, nil
+}
+
+func resolveGitTreeForPath(ctx context.Context, commitResolver *graphqlbackend.GitCommitResolver, paths pathsSet) (resolversByPathMap, error) {
```

Comment:
Plural to match the args

```suggestion
func resolveGitTreesForPaths(ctx context.Context, commitResolver *graphqlbackend.GitCommitResolver, paths pathsSet) (resolversByPathMap, error) {
```

Reply:
Updated all occurrences.